Custom ct duplicates

hanna-liashchuk opened the pull request:
resolves #

<!---
  Include the number of the issue addressed by this PR above if applicable.
-->

### Description

<!---
  Describe the Pull Request here. Add any references and info to help reviewers
  understand your changes. Include any tradeoffs you considered.
-->

### Checklist

- [ ] I have run this code in development and it appears to resolve the stated issue
- [ ] This PR includes tests, or tests are not required/relevant for this PR
- [ ] I have updated the README.md file

hanna-liashchuk committed:
Add functionality to delete orphaned rows in child tables (custom ct)

hanna-liashchuk committed:
fix

hanna-liashchuk committed:
fix

VladOrlov commented on src/main/scala/mirroring/services/writer/ChangeTrackingService.scala:
```diff
@@ -99,4 +111,50 @@ class ChangeTrackingService(
+
+    for (name <- deleteOperations.columns) {
+      deleteOperations = deleteOperations.withColumnRenamed(name, name.replace(sourceColPrefix, ""))
+    }
```
Maybe make sense to use `new Column("someColumn").as("someAlias")` in this case?

VShkaberda replied:
Removed completely.

VladOrlov commented on src/main/scala/mirroring/config/Config.scala:
```diff
@@ -145,6 +147,11 @@ case class Config(
   )
 
+  require(
+    CTChangesQuery.nonEmpty ^ parent_key.length == 0,
```
Shouldn't this predicate use && ? As I can understand from message below both conditions must be true...

VShkaberda replied:
Nope, it's fine, it requires whether `True, True` or `False, False`.

VShkaberda replied:
Require message has been made more clear.

VShkaberda committed:
Fix parent_key missed mapArgsDefault and make clear "reqiure" message.

VShkaberda committed:
Fix delete condition for object mirroring.

VShkaberda committed:
Distinguish change_tracking and object_mirroring.

ALtered flow for object_mirroring in ChangeTrackingService breaks default change_tracking.

VShkaberda committed:
Rename arg object_mirroring to custom_ct.

VShkaberda committed:
Add ct_window.

VShkaberda committed:
Add logging repartition.

VShkaberda committed:
Fix NULL ctDeltaVersion.

VShkaberda committed:
Override val logger for ChangeTrackingService.

VShkaberda committed:
Fix java.io.NotSerializableException.

VShkaberda committed:
Add readme.

VShkaberda committed:
Fix deprecated name in Config.

VladOrlov commented on src/main/scala/mirroring/services/MirroringManager.scala:
```diff
@@ -91,7 +97,9 @@ class MirroringManager {
     val jdbcDF: DataFrame =
-      if (config.isChangeTrackingEnabled && isDeltaTableExists && config.CTChangesQuery.nonEmpty) {
+      if (
+        (config.isChangeTrackingEnabled || config.isCustomCTEnabled) && isDeltaTableExists && config.CTChangesQuery.nonEmpty
```
I have suggestion to use one approach to namings - isChangeTrackingEnabled & isCustomChangeTrackingEnabled
or  isCTEnabled & isCustomCTEnabled. I prefer full names.

VShkaberda replied:
Done. Altered into the full name.

VladOrlov commented on src/main/scala/mirroring/services/databases/JdbcCTService.scala:
```diff
@@ -48,7 +48,7 @@ class JdbcCTService(jdbcContext: JdbcContext) extends Serializable {
       log.debug(schema)
       log.info("Executing procedure to create rdd...")
-      val myRDD: JavaRDD[Array[Object]] = JdbcRDD.create(
+      var myRDD: JavaRDD[Array[Object]] = JdbcRDD.create(
```
Using vars is bad practice

VShkaberda replied:
I think it may be refactored as
```scala
      val myRDD: JavaRDD[Array[Object]] = JdbcRDD
        .create(
          getSparkSession.sparkContext,
          connectionManager,
          jdbcContext.ctChangesQuery,
          params(0).toLong,
          params(1).toLong,
          1,
          r => JdbcRDD.resultSetToObjectArray(r)
        )
        .repartition(getSparkSession.conf.get("spark.sql.shuffle.partitions").toInt)
```
Let's refactor it in the next iteration,

hanna-liashchuk commented on src/main/scala/mirroring/config/Config.scala:
```diff
@@ -141,8 +145,20 @@ case class Config(
+    s"Parameter `primary_key` should be specified if `change_tracking` or `custom_ct` is true."
+  )
+
+  require(
```
couldn't it be written without if/else?

VShkaberda replied:
Changed.

VShkaberda committed:
Remove if in require.

hanna-liashchuk commented:
not allowed to approve, lgtm

VShkaberda committed:
Alter CustomChangeTrackingService write flow for an empty dataset.

hanna-liashchuk commented on build.sbt:
```diff
@@ -37,7 +37,7 @@ scalaVersion := "2.12.10"
 val sparkVersion    = "3.3.2"
 val airframeVersion = "23.6.1"
 val deltaVersion    = "2.2.0"
-version := s"1.1.6-spark${sparkVersion}-scala${scalaVersion.value}"
```
do we really want to merge this version?

VShkaberda replied:
Restored version.

hanna-liashchuk commented on src/main/scala/mirroring/services/writer/CustomChangeTrackingService.scala:
```diff
@@ -49,33 +49,44 @@ class CustomChangeTrackingService(
+  private def mergeData(data: DataFrame): Unit = {
+    verifySchemaMatch(data)
 
+    val updOperations = data.as(Config.SourceAlias).where(insertCondition)
```
do we really need to make things so complicated? what if there are no delete operations? will it still read the entire table to make a commit?

VShkaberda replied:
If there are no delete operation, `deleteRows` with executing merge leads to SnaphotIsolation commit without reading the whole table.

hanna-liashchuk replied:
ok then

hanna-liashchuk commented:
lgtm, except for the version

VShkaberda committed:
Restore version.

VShkaberda committed:
Fix CustomChangeTrackingService logger name.

VShkaberda committed:
Add TO DO for refactoring.

VShkaberda committed:
Alter isCustomCTEnabled into isCustomChangeTrackingEnabled.
